drivers/lsm6ds3/lsm6ds3.c

```diff
@@ -0,0 +1,296 @@
+    data->y_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->acc_address,
+                        LSM6DS3_REG_OUT_Y_H_XL, &tmp);
+    data->y_axis |= tmp<<8;
+    res += i2c_read_reg(dev->i2c, dev->acc_address,
+                        LSM6DS3_REG_OUT_Z_L_XL, &tmp);
+    data->z_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->acc_address,
+                        LSM6DS3_REG_OUT_Z_H_XL, &tmp);
+    data->z_axis |= tmp<<8;
+    res += i2c_write_reg(dev->i2c, dev->acc_address,
+                         LSM6DS3_REG_CTRL3_C, ~(1<<7));
+    i2c_release(dev->i2c);
+
+    data->x_axis = data->x_axis>>4;
+    data->y_axis = data->y_axis>>4;
+    data->z_axis = data->z_axis>>4;
+
+    if (res != 6)
+    {
+		    DEBUG("[ FAIL ]\n");
+		    return -1;
+    }
+
+    DEBUG("[ OK ]\n");
+    return 0;
+}
+
+int lsm6ds3_read_gyr(lsm6ds3_t *dev, lsm6ds3_3d_data_t *data)
+{
+	  int res = 0;
+    char tmp;
+
+	  i2c_acquire(dev->i2c);
+    i2c_read_reg(dev->12c, dev-acc_address, LSM6DS3_STATUS_REG, &tmp);
+    DEBUG("LSM6DS3: status ... %x\n", tmp);
+    res += i2c_write_reg(dev->i2c, dev->gyr_address,
+                         LSM6DS3_REG_CTRL3_C, LSM6DS3_CTRL3_C_BDU);
+    DEBUG("LSM6DS3: retrieve gyroscope values ... ");
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_X_L_G, &tmp);
+    data->x_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_X_H_G, &tmp);
+    data->x_axis = tmp<<8;
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_Y_L_G, &tmp);
+    data->y_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_Y_H_G, &tmp);
+    data->y_axis = tmp<<8;
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_Z_L_G, &tmp);
+    data->z_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->gyr_address,
+                        LSM6DS3_REG_OUT_Z_H_G, &tmp);
+    data->z_axis = tmp<<8;
+    res += i2c_write_reg(dev->i2c, dev->gyr_address,
+                         LSM6DS3_REG_CTRL3_C, ~(1<<7));
+	  i2c_release(dev->i2c);
+
+    data->x_axis = data->x_axis>>4;
+    data->y_axis = data->y_axis>>4;
+    data->z_axis = data->z_axis>>4;
+
+	  if (res != 8)
+	  {
+		    DEBUG("[ FAIL ]\n");
+		    return -1;
+	  }
+
+	  DEBUG("[ OK ]\n");
+  	return 0;
+}
+
+int lsm6ds3_read_temp(lsm6ds3_t *dev, int16_t *value)
+{
+	  int res = 0;
+    char tmp;
+
+	  i2c_acquire(dev->i2c);
+    i2c_read_reg(dev->12c, dev-acc_address, LSM6DS3_STATUS_REG, &tmp);
+    DEBUG("LSM6DS3: status ... %x\n", tmp);
+    res += i2c_write_reg(dev->i2c, dev->acc_address,
+                         LSM6DS3_REG_CTRL3_C, LSM6DS3_CTRL3_C_BDU);
+    DEBUG("LSM6DS3: retrieve temperature data ... ");
+    res += i2c_read_reg(dev->i2c, dev->acc_address,
+                        LSM6DS3_OUT_TEMP_L, &tmp);
+    *value += tmp;
```

Comment:
`+=`? I think I would prepare the pointer at the end and spend a temporary integer variable or two chars. If this works (we'll see when the hardware is ready) it's up to you.